### ISSUE TITLE
coturn external-ip from DNS?

### ISSUE THREAD

**Issue by tfreedman** [coturn/coturn#563]
I have a coturn server running on my NAT, but do not have a static IP address from my ISP. However, I have DDNS set up for all other network services, such that my domain name always points to my current IP address. Given the way coturn's config file works, I would have to script ddclient to rewrite coturn's config file with my current IP address and restart coturn to use the new config each time the address changes. It would be much easier if I could just supply a domain name for coturn to check the A record of in place of external-ip.

**Comment by misi** [comment]
Sorry for my late reply! @tfreedman 
I am happy to merge a Pull Request if you write a fix for dns resolution, but I feel it is simple to make it in scripts.

**Comment by foresto** [comment]
@tfreedman Have you tried setting external-ip (or -X) to your host's domain name instead of an address? I just did, and it seems to work.

When mainrelay.c parses that option, it gets sent through `make_ioa_addr()`, [here](https://github.com/coturn/coturn/blob/9b1cca1fbe909e7cc7c7ac28865f9c190af3515b/src/apps/relay/mainrelay.c#L1379), which then passes it to `getaddrinfo()`, [here](https://github.com/coturn/coturn/blob/9b1cca1fbe909e7cc7c7ac28865f9c190af3515b/src/client/ns_turn_ioaddr.c#L234), which resolves the domain name to an IP address. (At least, that's what I gathered from a superficial look at the code; perhaps @misi can verify this.)

So, I think all that's really needed here is documenting this behavior, and probably removing the "For Amazon EC2 users" heading in the example config file.

Of course, if your IP address changes, you might have to restart coturn in order for it to resolve the domain name again.

**Comment by eakraly** [comment]
Hi @tfreedman and @foresto 
Though it seems that it works I would strongly advise against relying on turnserver resolving an address
This adds unrelated functionality to the code - same can be achieved in many different ways for different scenario while passing `--external-ip` as argument or updating config file.

What's missing for config file is support for reloading of configuration file without restarting the process #878

For now I suggest to close this issue (since it's been 1.5 years since last update)

**Comment by foresto** [comment]
Thanks for the link to #878. It wouldn't solve the problem on its own, though, as it would still require some additional process to monitor the external IP address and babysit coturn. I think people with humble server needs would appreciate not having to deal with the extra complexity.

**Comment by eakraly** [comment]
@foresto I do not see what the issue is - is it that you expect turnserver to refresh its external IP periodically without external mechanisms?
This is not something that can be "fixed" - this is a new feature that does not exist yet AFAIK

Is this something that you can submit a PR for?

**Comment by foresto** [comment]
Not right now, but if it's still open next time I have the need, I might.

**Comment by eakraly** [comment]
Thank you @foresto 
I was thinking your issue was addressed based on the fact that -X resolves the address (but missed the fact that there is no refresh mechanism)
Marking this issue as enhancement (instead of question)

**Comment by coturn-codeveloper6092** [comment]
I have pushed the idea of Coturn resolving its host domain name to the limit: [I have added](https://github.com/coturn/coturn/compare/master...coturn-codeveloper6092:coturn:dynamic-dns) the option `--external-hostname <hostname>/<private-ip>` which is similar to `--external-ip` but `<hostname>` is resolved every time when Coturn needs to know its public IP address.

**Comment by j007bond007** [comment]
Also wondering when this will be merged, thankyou!

**Comment by tfreedman** [comment]
Just hit this bug again, since I had to make a call earlier today via Matrix/Synapse which relies on coturn.

Scripting this isn't a good workaround, as it would require restarting coturn whenever the external IP address changes. This is akin to having to restart a web server if my modem's external IP address changes, even though the web server will always be capable of receiving traffic regardless of what external address it has. You'd also have to poll the external IP constantly, and there would ultimately be a window of time that coturn doesn't work because the external IP address has changed but we haven't polled for an update yet.

The solution by @coturn-codeveloper6092 looks like it would do the trick, by freshly loading the IP address whenever it's actually required. Are there any plans to actually merge it?

**Comment by eakraly** [comment]
@tfreedman 
>This is akin to having to restart a web server if my modem's external IP address changes, even though the web server will always be capable of receiving traffic regardless of what external address it has

You are wrong here. If your external address change your webserver will NOT be able to receive any existing traffic because it will be routed to some other place (you have a different IP now) until you update DNS. At this point your traffic is 0.
Moreover, TURN is not like webserver - it needs to know its external IP to operate.

Change in IP needs to be detected and handled - you just want turnserver to do it. The issue is that many different scenarios exist to how IPs change and what needs to be done about it. Proposed solution is not really TURN related and only complicates coturn code for a very limited range of scenarios

**Comment by OdinVex** [comment]
The method for obtaining an external address via DNS has another unfortunate downside, restricted DNS environments. Our environment blocks DNS to any provider, systems must use our own in-house DNS servers, so it's broken for now.

**Comment by kralge** [comment]
It is quite disappointing that the problem many of us facing even has a solution, but I need to start hacking around because it does not even seem to be intended to be merged at all.

If turnserver can solve this problem then it should give the user the option to do so. Every new feature complicates the code and this should not be a problem up until it means no penalty to those how are unaffected. This is how a software becomes versatile and usable for the widest range of scenarios.

So please consider merging this option.
